pow/src/main/java/tech/pegasys/artemis/pow/DepositRequestManager.java

```diff
@@ -0,0 +1,222 @@
+  private BigInteger latestSuccessfullyQueriedBlockNumber = BigInteger.ZERO;
+
+  public DepositRequestManager(
+      Eth1Provider eth1Provider,
+      AsyncRunner asyncRunner,
+      DepositEventChannel depositEventChannel,
+      DepositContract depositContract) {
+    this.eth1Provider = eth1Provider;
+    this.asyncRunner = asyncRunner;
+    this.depositEventChannel = depositEventChannel;
+    this.depositContract = depositContract;
+  }
+
+  public void start() {
+    newBlockSubscription =
+        eth1Provider
+            .getLatestBlockFlowable()
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .subscribe(this::updateLatestCanonicalBlockNumber, this::onSubscriptionFailed);
+  }
+
+  private void onSubscriptionFailed(Throwable err) {
+    Disposable subscription = newBlockSubscription;
+    if (subscription != null) {
+      subscription.dispose();
+    }
+    LOG.warn("New block subscription failed, retrying.", err);
+    start();
+  }
+
+  public void stop() {
+    newBlockSubscription.dispose();
+  }
+
+  private synchronized void updateLatestCanonicalBlockNumber(
+      BigInteger latestCanonicalBlockNumber) {
+    this.latestCanonicalBlockNumber = latestCanonicalBlockNumber;
+    getLatestDeposits();
+  }
+
+  private void getLatestDeposits() {
+    final BigInteger latestCanonicalBlockNumberAtRequestStart;
+    final BigInteger latestSuccessfullyQueriedBlockNumberAtRequestStart;
+    synchronized (DepositRequestManager.this) {
+      if (active) {
+        requestQueued = true;
+        return;
+      } else {
+        active = true;
+        requestQueued = false;
+      }
+
+      latestCanonicalBlockNumberAtRequestStart = this.latestCanonicalBlockNumber;
+      latestSuccessfullyQueriedBlockNumberAtRequestStart =
+          this.latestSuccessfullyQueriedBlockNumber;
+    }
+
+    LOG.trace(
+        "Attempting to fetch deposit events for block numbers in the range ({}, {})",
+        latestSuccessfullyQueriedBlockNumberAtRequestStart,
+        latestCanonicalBlockNumberAtRequestStart);
+
+    DefaultBlockParameter fromBlock =
+        DefaultBlockParameter.valueOf(latestSuccessfullyQueriedBlockNumberAtRequestStart);
+    DefaultBlockParameter toBlock =
+        DefaultBlockParameter.valueOf(latestCanonicalBlockNumberAtRequestStart);
+
+    fetchAndPublishDepositEventsInBlockRange(fromBlock, toBlock)
+        .finish(
+            () -> onDepositRequestSuccessful(latestCanonicalBlockNumberAtRequestStart),
+            (err) -> onDepositRequestFailed(err, latestCanonicalBlockNumberAtRequestStart));
+  }
+
+  private synchronized void onDepositRequestSuccessful(
+      BigInteger latestCanonicalBlockNumberAtRequestStart) {
+    this.latestSuccessfullyQueriedBlockNumber = latestCanonicalBlockNumberAtRequestStart;
+    active = false;
+    if (requestQueued) {
+      getLatestDeposits();
+    }
```

Comment:
It occurs to me that we can remove the `requestQueued` variable and just use `if (latestCanonicalBlockNumber.compareTo(latestSuccessfullyQueriedBlockNumber) > 0)` here. The places where we set `requestQueued` can just be removed.

Basically, if we haven't reached the latest canonical block, we need another request.

Reply:
That seems like a no brainer now in hindsight. Thanks for the detailed review.

Reply:
Done.